### PR TITLE
Add persistence enable-disable to wide_ip

### PR DESCRIPTION
**Pull request by nitinthewiz**
Adds missing persistence and related persist_cidr_ipv4, persist_cidr_ipv6, and persistence_ttl options to the wide_ip

### CODE REVIEW

**Comment by wojtek0806** [comment]
once this PR is merged do full rebase and let the tests pass before we move on to this:

https://github.com/F5Networks/f5-ansible/pull/1540

**Comment by nitinthewiz** [comment]
The build failed due to the following tests. This seems to be a part of the rebase, since my changes passed before the rebase. - 

ERROR: The 2 sanity test(s) listed below (out of 31) failed. See error output above for details.
action-plugin-docs
ignores

**Comment by wojtek0806** [comment]
i need to see why do we get this error in shippable going to try a different branch

File: library/modules/bigip_gtm_wide_ip.py [hunk 1]
```diff
@@ -119,6 +119,42 @@
...
+      - disabled
+      - enabled
+    default: disabled
+    version_added: 2.9
+  persistence_ttl:
+    description:
+      - Specifies the time to maintain a connection between an local DNS and
+        a particular virtual server. The default setting is 3600 seconds.
+    type: int
+    default: 3600
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: no defaults

File: library/modules/bigip_gtm_wide_ip.py [hunk 2]
```diff
@@ -119,6 +119,42 @@
...
+    version_added: 2.9
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: incorrect version, 2.9 was already released

File: library/modules/bigip_gtm_wide_ip.py [hunk 3]
```diff
@@ -119,6 +119,42 @@
...
+  persist_cidr_ipv4:
+    description:
+      - Specifies a mask used to group IPv4 LDNS addresses. This feature
+        allows one persistence record to be shared by LDNS addresses
+        that match within this mask. The default is 32.
+    type: int
+    default: 32
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: no defaults

File: library/modules/bigip_gtm_wide_ip.py [hunk 4]
```diff
@@ -119,6 +119,42 @@
...
+    version_added: 2.9
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: incorrect version

File: library/modules/bigip_gtm_wide_ip.py [hunk 5]
```diff
@@ -119,6 +119,42 @@
...
+      - Specifies the time to maintain a connection between an local DNS and
+        a particular virtual server. The default setting is 3600 seconds.
+    type: int
+    default: 3600
+    version_added: 2.9
+  persist_cidr_ipv4:
+    description:
+      - Specifies a mask used to group IPv4 LDNS addresses. This feature
+        allows one persistence record to be shared by LDNS addresses
+        that match within this mask. The default is 32.
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: we do not inform what the default is by device as it may change version to version

File: library/modules/bigip_gtm_wide_ip.py [hunk 6]
```diff
@@ -119,6 +119,42 @@
...
+    type: str
+    choices:
+      - disabled
+      - enabled
+    default: disabled
+    version_added: 2.9
+  persistence_ttl:
+    description:
+      - Specifies the time to maintain a connection between an local DNS and
+        a particular virtual server. The default setting is 3600 seconds.
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: we do not inform what the default is by device as it may change version to version

File: library/modules/bigip_gtm_wide_ip.py [hunk 7]
```diff
@@ -119,6 +119,42 @@
...
+  persistence:
+    description:
+      - When C(enabled), ensures that when a local DNS makes repetitive requests on
+        behalf of a client, the system reconnects the client to the same resource
+        as previous requests.
+      - When C(disabled), ensures repetitive requests do not reconnect the client
+        to the same resource.
+    type: str
+    choices:
+      - disabled
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: this should be boolean, change it to boolean, look at other modules how to do this

File: library/modules/bigip_gtm_wide_ip.py [hunk 8]
```diff
@@ -209,6 +261,11 @@
   returned: changed
   type: list
   sample: ['alias1.foo.com', '*.wildcard.domain']
+persistence:
+  description: Whether the pool connections will be persisted.
+  returned: changed
+  type: str
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: change to boolean

File: library/modules/bigip_gtm_wide_ip.py [hunk 9]
```diff
@@ -119,6 +119,42 @@
...
+    type: int
+    default: 32
+    version_added: 2.9
+  persist_cidr_ipv6:
+    description:
+      - Specifies a mask used to group IPv6 LDNS addresses. This feature
+        allows one persistence record to be shared by LDNS addresses
+        that match within this mask. The default is 128.
+    type: int
+    default: 128
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: no defaults

File: library/modules/bigip_gtm_wide_ip.py [hunk 10]
```diff
@@ -119,6 +119,42 @@
...
+        allows one persistence record to be shared by LDNS addresses
+        that match within this mask. The default is 32.
+    type: int
+    default: 32
+    version_added: 2.9
+  persist_cidr_ipv6:
+    description:
+      - Specifies a mask used to group IPv6 LDNS addresses. This feature
+        allows one persistence record to be shared by LDNS addresses
+        that match within this mask. The default is 128.
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: same as above

File: library/modules/bigip_gtm_wide_ip.py [hunk 11]
```diff
@@ -119,6 +119,42 @@
...
+    type: int
+    default: 128
+    version_added: 2.9
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: wrong version

File: library/modules/bigip_gtm_wide_ip.py [hunk 12]
```diff
@@ -922,6 +1017,22 @@ def __init__(self):
                 type='list'
             ),
             last_resort_pool=dict(),
+            persistence=dict(
+                default='disabled',
+                choices=['enabled', 'disabled']
+            ),
+            persistence_ttl=dict(
+                type='int',
+                default=3600
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: remove defaults

File: library/modules/bigip_gtm_wide_ip.py [hunk 13]
```diff
@@ -922,6 +1017,22 @@ def __init__(self):
                 type='list'
             ),
             last_resort_pool=dict(),
+            persistence=dict(
+                default='disabled',
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: no defaults

File: library/modules/bigip_gtm_wide_ip.py [hunk 14]
```diff
@@ -922,6 +1017,22 @@ def __init__(self):
...
+                choices=['enabled', 'disabled']
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: should be boolean

File: library/modules/bigip_gtm_wide_ip.py [hunk 15]
```diff
@@ -922,6 +1017,22 @@ def __init__(self):
...
+                type='int',
+                default=3600
+            ),
+            persist_cidr_ipv4=dict(
+                type='int',
+                default=32
+            ),
+            persist_cidr_ipv6=dict(
+                type='int',
+                default=128
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: remove

File: library/modules/bigip_gtm_wide_ip.py [hunk 16]
```diff
@@ -922,6 +1017,22 @@ def __init__(self):
...
+                default='disabled',
+                choices=['enabled', 'disabled']
+            ),
+            persistence_ttl=dict(
+                type='int',
+                default=3600
+            ),
+            persist_cidr_ipv4=dict(
+                type='int',
+                default=32
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: remove

File: library/modules/bigip_gtm_wide_ip.py [hunk 17]
```diff
@@ -409,6 +480,30 @@ def aliases(self):
         self._values['aliases'].sort()
         return self._values['aliases']
 
+    @property
+    def persistence(self):
+        if self._values['persistence'] is None:
+            return None
+        return str(self._values['persistence'])
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: change

File: library/modules/bigip_gtm_wide_ip.py [hunk 18]
```diff
@@ -409,6 +480,30 @@ def aliases(self):
...
+
+    @property
+    def persistence_ttl(self):
+        if self._values['persistence_ttl'] is None:
+            return None
+        return self._values['persistence_ttl']
+
+    @property
+    def persist_cidr_ipv4(self):
+        if self._values['persist_cidr_ipv4'] is None:
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why is this here? only use property when you are actually checking or doing value transformation. Also if you have int values usually its best to check if the number is valid, check the device documentation for valid range, as far as implementation goes you need  to look at other modules for established patterns use patterns

File: library/modules/bigip_gtm_wide_ip.py [hunk 19]
```diff
@@ -409,6 +480,30 @@ def aliases(self):
...
+            return None
+        return self._values['persist_cidr_ipv4']
+
+    @property
+    def persist_cidr_ipv6(self):
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why is this here? only use property when you are actually checking or doing value transformation. Also if you have int values usually its best to check if the number is valid, check the device documentation for valid range, as far as implementation goes you need to look at other modules for established patterns use patterns

File: library/modules/bigip_gtm_wide_ip.py [hunk 20]
```diff
@@ -409,6 +480,30 @@ def aliases(self):
...
+    @property
+    def persistence(self):
+        if self._values['persistence'] is None:
+            return None
+        return str(self._values['persistence'])
+
+    @property
+    def persistence_ttl(self):
+        if self._values['persistence_ttl'] is None:
+            return None
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why is this here? only use property when you are actually checking or doing value transformation. Also if you have int values usually its best to check if the number is valid, check the device documentation for valid range, as far as implementation goes you need to look at other modules for established patterns use patterns

File: library/modules/bigip_gtm_wide_ip.py [hunk 21]
```diff
@@ -209,6 +261,11 @@
   returned: changed
   type: list
   sample: ['alias1.foo.com', '*.wildcard.domain']
+persistence:
+  description: Whether the pool connections will be persisted.
+  returned: changed
+  type: str
+  sample: disabled
 '''
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: where are the rest of returned values from returnables, document every value the module returns after it runs

**Comment by wojtek0806** [comment]
Also no need to have provider/env type tests, just do the env type tests, some older modules had it due to some issues with arg.spec which has been removed. Also not having these tests means less work for us once we start tidying up those tests.

**Comment by nitinthewiz** [comment]
> Also no need to have provider/env type tests, just do the env type tests, some older modules had it due to some issues with arg.spec which has been removed. Also not having these tests means less work for us once we start tidying up those tests.

I took a call because there was no documentation on which type of test is older. I'll just put my test outside of these.

**Comment by wojtek0806** [comment]
all you need to do is look at the latest PRs all of them do not use this distinction

**Comment by nitinthewiz** [comment]
I checked and the file in provider was there by mistake. The actual test is only under env.

File: library/modules/bigip_gtm_wide_ip.py (hunk 5 above)
Comment (nitinthewiz) [reply library/modules/bigip_gtm_wide_ip.py]: There are other scenarios, many of them, where the default has been specified. Removing regardless.

File: library/modules/bigip_gtm_wide_ip.py (hunk 7 above)
Comment (nitinthewiz) [reply library/modules/bigip_gtm_wide_ip.py]: changed. I was following the module's current pattern.

File: library/modules/bigip_gtm_wide_ip.py (hunk 18 above)
Comment (nitinthewiz) [reply library/modules/bigip_gtm_wide_ip.py]: added checking.

File: library/modules/bigip_gtm_wide_ip.py (hunk 19 above)
Comment (nitinthewiz) [reply library/modules/bigip_gtm_wide_ip.py]: added checking.

File: library/modules/bigip_gtm_wide_ip.py (hunk 20 above)
Comment (nitinthewiz) [reply library/modules/bigip_gtm_wide_ip.py]: added checking.

**Comment by nitinthewiz** [comment]
Ok, shippable failed as expected and Travis passed.

File: library/modules/bigip_gtm_wide_ip.py [hunk 22]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
         self._values['aliases'].sort()
         return self._values['aliases']
 
+    @property
+    def persistence(self):
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: This is incorrect, this is how you should do the boolean:

https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_profile_http.py#L1035

File: library/modules/bigip_gtm_wide_ip.py [hunk 23]
```diff
@@ -453,6 +572,15 @@ def last_resort_pool(self):
             return 'none'
         return self._values['last_resort_pool'].split(' ')[1]
 
+    @property
+    def persistence(self):
+        if self._values['persistence'] == 'enabled':
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why is this here? if API accepts enabled/disabled then this is not needed

File: library/modules/bigip_gtm_wide_ip.py [hunk 24]
```diff
@@ -119,6 +119,31 @@
         module.
     type: str
     version_added: 2.8
+  persistence:
+    description:
+      - When C(enabled), ensures that when a local DNS makes repetitive requests on
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: booleans always flatten to yes/no not enabled/disabled

File: library/modules/bigip_gtm_wide_ip.py [hunk 25]
```diff
@@ -922,6 +1050,10 @@ def __init__(self):
                 type='list'
             ),
             last_resort_pool=dict(),
+            persistence=dict(type='bool', default=False),
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: no defaults

File: library/modules/bigip_gtm_wide_ip.py [hunk 26]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
...
+            "Valid 'persist_cidr_ipv4' must be in range 0 - 4294967295."
+        )
+
+    @property
+    def persist_cidr_ipv6(self):
+        if not self._values['persistence']:
+            return 128
+        if self._values['persist_cidr_ipv6'] is None:
+            return None
+        if 0 <= int(self._values['persist_cidr_ipv6']) <= 4294967295:
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: if this is an int why do you cast the value to an int?

File: library/modules/bigip_gtm_wide_ip.py [hunk 27]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
...
+        if 0 <= int(self._values['persist_cidr_ipv4']) <= 4294967295:
+            return int(self._values['persist_cidr_ipv4'])
+        raise F5ModuleError(
+            "Valid 'persist_cidr_ipv4' must be in range 0 - 4294967295."
+        )
+
+    @property
+    def persist_cidr_ipv6(self):
+        if not self._values['persistence']:
+            return 128
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why do you return values if they are not declared by the user? remove this

File: library/modules/bigip_gtm_wide_ip.py [hunk 28]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
...
+            "Valid 'persistence_ttl' must be in range 0 - 4294967295."
+        )
+
+    @property
+    def persist_cidr_ipv4(self):
+        if not self._values['persistence']:
+            return 32
+        if self._values['persist_cidr_ipv4'] is None:
+            return None
+        if 0 <= int(self._values['persist_cidr_ipv4']) <= 4294967295:
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: if this is an int why do you cast the value to an int?

File: library/modules/bigip_gtm_wide_ip.py [hunk 29]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
...
+        if 0 <= int(self._values['persistence_ttl']) <= 4294967295:
+            return int(self._values['persistence_ttl'])
+        raise F5ModuleError(
+            "Valid 'persistence_ttl' must be in range 0 - 4294967295."
+        )
+
+    @property
+    def persist_cidr_ipv4(self):
+        if not self._values['persistence']:
+            return 32
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why do you return values if they are not declared by the user? remove this

File: library/modules/bigip_gtm_wide_ip.py [hunk 30]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
...
+    def persistence(self):
+        if self._values['persistence'] is None:
+            return None
+        if self._values['persistence']:
+            return 'enabled'
+        return 'disabled'
+
+    @property
+    def persistence_ttl(self):
+        if not self._values['persistence']:
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: why do you return values if they are not declared by the user? remove this

File: library/modules/bigip_gtm_wide_ip.py [hunk 31]
```diff
@@ -409,6 +484,50 @@ def aliases(self):
...
+            return 3600
+        if self._values['persistence_ttl'] is None:
+            return None
+        if 0 <= int(self._values['persistence_ttl']) <= 4294967295:
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: if this is an int why do you cast the value to an int?

File: library/modules/bigip_gtm_wide_ip.py [hunk 32]
```diff
@@ -186,6 +211,22 @@
...
       server: lb.mydomain.com
   delegate_to: localhost
+
+- name: Assign a pool with persistence to the Wide IP
+  bigip_gtm_wide_ip:
+    pool_lb_method: round-robin
+    name: my-wide-ip.example.com
+    pools:
+      - name: pool1
+        persistence: True
```
Comment (wojtek0806) [inline library/modules/bigip_gtm_wide_ip.py]: change

File: test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+
+- name: Issue Persistence - Create wide IP without persistence
+  bigip_gtm_wide_ip:
+    lb_method: "{{ pool_lb_method }}"
+    state: present
+    type: a
+    name: "{{ wide_ip2 }}"
+    pools:
+      - name: "{{ pool_name }}"
+  when: f.system_info.product_version  >= "12.0.0"
```
Comment (wojtek0806) [inline test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml]: this kind of check needs to be done before the test is run

File: test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+- name: Issue Persistence - Provision GTM
+  bigip_provision:
+    module: gtm
+  tags:
+    - module-provisioning
+
+- name: Issue Persistence - Collect BIG-IP facts
+  bigip_device_facts:
+    include: system-info
+  register: f
```
Comment (wojtek0806) [inline test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml]: move it to main.yml and apply the version check to the entire task import, also gtm support for v11 is deprecated so it is not really needed

File: library/modules/bigip_gtm_wide_ip.py (hunk 22 above)
Comment (nitinthewiz) [reply library/modules/bigip_gtm_wide_ip.py]: Oh that's an interesting function for tackling boolean! I'll use it!

File: test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml (hunk 33 above)
Comment (nitinthewiz) [reply test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml]: I'm just following one of the recent standards here - https://github.com/F5Networks/f5-ansible/blob/ba8e3002abe6d3075a611c7dc0427bb77ccb2281/test/integration/targets/bigip_network_globals/tasks/main.yaml

File: test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml (hunk 34 above)
Comment (nitinthewiz) [reply test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml]: removed.

File: test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml (hunk 33 above)
Comment (wojtek0806) [reply test/integration/targets/bigip_gtm_wide_ip/tasks/environment/issue-persistence.yaml]: this would make sense if you had a supported version in mind and some tasks were supposed to run on both versions, since you want all to run on v12 and beyond then it is not needed, also  v11 is deprecated so we can get rid of this :)

**Comment by nitinthewiz** [comment]
I've deleted the product version check, as you recommended.
